### PR TITLE
'Fix' hang in ScriptPromise tests

### PR DESCRIPTION
**Pull request by cbaggers**
Previous there were 3 promise tests sharing the same test panel.

the panel has a javascript tag

the javascript tag would, at module eval time, try and dispatch some
promises.

.e.g this top level expression `customPanel.getString().then(...)`

however the test only populated the promise it was interested in, so
I believe the other expressions would try and call `then` on null.

This seems to leave the js in a rather undefined that so that a few
times in a hundred the promise never resolves and we get a hang
resulting in the tests timing out.

By refactoring these tests to each own their own panel and
bootstrapping we avoid the issue.

However the lack of an error is a deeper issue that is being tracked
here: https://github.com/fusetools/fuselibs-public/issues/840

** Briefly describe changes and the motivation behind them here **

This PR contains:
- [ ] Changelog
- [ ] Documentation
- [x] Tests

### CODE REVIEW

**Comment by kusma** [comment]
I'm not a big fan of this approach. The reason is that we end up duplicating code. All we would have needed would be to fill in these promises with some dummies, no?

**Comment by cbaggers** [comment]
@kusma Sounds fair. I'll try that now

**Comment by cbaggers** [comment]
@kusma I think my theory of what was wrong may be wrong. I tried both 'set the promises at the start of each test' approach e.g.
```
 		{
 			var child = new UX.BasicTest();
 			child.customPanel.StringFuture = new Promise<string>("this is a string");
+			child.customPanel.ObjectFuture = new Promise<TestObject>(new TestObject());
+			child.customPanel.NumberFuture = new Promise<double>(0.0);
```
and also the 'give the properties defaults' approach
```
-		public Future<string> StringFuture { get; set; }
-		public Future<double> NumberFuture { get; set; }
-		public Future<TestObject> ObjectFuture { get; set; }
+		Future<string> _stringFuture = new Promise<string>("ignored promise");
+		Future<double> _numberFuture = new Promise<double>(0.0);
+		Future<TestObject> _objectFuture = new Promise<TestObject>(new TestObject());
+
+		public Future<string> StringFuture
+		{
+			get { return _stringFuture; }
+			set { _stringFuture = value; }
+		}
+		public Future<double> NumberFuture
+		{
+			get { return _numberFuture; }
+			set { _numberFuture = value; }
+		}
+		public Future<TestObject> ObjectFuture
+		{
+			get { return _objectFuture; }
+			set { _objectFuture = value; }
+		}
```

and both still resulting in the fail test hanging fairly frequently.

However the approach in this pr still works. Very odd

**Comment by mortoray** [comment]
You can probably simplify this by using `FuseTest.DudElement`. It lets you capture bindings by various types without creating a custom class.  Often the ux:Property's can interfere with what you're actually trying to test.

File: Source/Fuse.Scripting/Tests/ScriptClass/ScriptStringPromise.Test.uno [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	public class ScriptStringPromiseTest : TestBase
+	{
+		[Test]
+		public void StringTest()
+		{
+			var child = new UX.StringBasicTest();
+			child.customStringPanel.StringFuture = new Promise<string>("this is a string");
+			using (var root = TestRootPanel.CreateWithChild(child))
+			{
+				while (string.IsNullOrEmpty(child.StringProperties.StringResult))
```
Comment (mortoray) [inline Source/Fuse.Scripting/Tests/ScriptClass/ScriptStringPromise.Test.uno]: I'd like to see a tighter requirement on this. I'm thinking a `MultiStepFrameJS(2)` would be correct here. We expect only the first JS to run and request the future, then the UI thread produce the value, then one more JS step to get the value out.

Comment (kusma) [reply Source/Fuse.Scripting/Tests/ScriptClass/ScriptStringPromise.Test.uno]: Yeah, this was my crappy attempt at making this more robust, but I think it's been shown that I misidentified the problem.

**Comment by mortoray** [comment]
@cbaggers Were you going to return to this PR or is it no longer relevant?

**Comment by mortoray** [comment]
I'll assume this is abandoned.

**Pull request closed without merging**